### PR TITLE
A library of custom match procs for testing

### PR DESCRIPTION
**Pull request by glennj**
Please see commentary in file.

### CODE REVIEW

**Comment by sshine** [comment]
I like that we collect these test matchers, since the framework appears insufficient for these and possibly others (we may want a "list as set" matcher, and so on). But I would like to consider the alternative to embed any necessary test matchers in the test suite file, also making the student's download self-contained.

*(Another alternative, of which I don't know the viability, is to keep a symlink in the GitHub repo so we only have one instance of this collection of test matchers. I don't think this is viable for two reasons: One is that I don't know if the file will be copied properly being a symlink (it might), and the other is that I don't know if the CLI will register that a change is made to an exercise's files when there technically, in the case of a symlink to a changed file, isn't.)*

---

In the Standard ML (SML) track, there is no package manager or standardized test framework, so we distribute a testlib.sml along with the test suite. I've had to update this test suite once (to enable floating-point comparisons with an epsilon in one exercise, since comparing floats for equality in SML is a type error), which meant that there were now two versions of testlib.sml, one for this particular exercise and one used everywhere else. Updating all exercises every time a change is made to testlib.sml seems like clutter, so for the SML track, the premise of diverging testlib.sml seems unavoidable and sadly acceptable.

I wonder if we can avoid something similar here by only distributing the exercise's test suite and by embedding the necessary custom match procs in that file. As we write a test suite generator, we can parameterise it with the kind of testing necessary and automatically embed any custom match procs that may be necessary. That way we only distribute one file, and updating or adding one test helper does not cause a divergence in all exercises that uses *any* test helper.

**Comment by sshine** [comment]
TL;DR: I will merge this file, but my long comment discusses alternative commentaries to the file.

**Comment by sshine** [comment]
Yes, I agree. Later we can automatically embed the necessary snippets if we generate tests.
